go/vt/vttablet/grpctmclient/client.go

```diff
@@ -218,44 +225,66 @@ func (client *grpcClient) dialDedicatedPool(ctx context.Context, dialPoolGroup D
 		return nil, nil, err
 	}
 
-	client.mu.Lock()
-	defer client.mu.Unlock()
+	client.dedicatedPoolMu.Lock()
+
 	if client.rpcDialPoolMap == nil {
 		client.rpcDialPoolMap = make(map[DialPoolGroup]addrTmcMap)
 	}
 	if _, ok := client.rpcDialPoolMap[dialPoolGroup]; !ok {
 		client.rpcDialPoolMap[dialPoolGroup] = make(addrTmcMap)
 	}
+
 	m := client.rpcDialPoolMap[dialPoolGroup]
-	if _, ok := m[addr]; !ok {
-		tm, err := client.createTmc(ctx, addr, opt)
-		if err != nil {
-			return nil, nil, err
-		}
-		m[addr] = tm
+	entry, ok := m[addr]
+	if !ok {
+		entry = &tmcEntry{}
+		m[addr] = entry
 	}
+	client.dedicatedPoolMu.Unlock()
```

Comment:
In my opinion, using an inline function helps making it easier to understand that the mutex is going to be unlocked correctly again, as locking and and the defer for the unlock are places right next to each other. What do you think?

```suggestion
	entry := func() *tmcEntry {
		client.dedicatedPoolMu.Lock()
		defer client.dedicatedPoolMu.Unlock()

		if client.rpcDialPoolMap == nil {
			client.rpcDialPoolMap = make(map[DialPoolGroup]addrTmcMap)
		}
		if _, ok := client.rpcDialPoolMap[dialPoolGroup]; !ok {
			client.rpcDialPoolMap[dialPoolGroup] = make(addrTmcMap)
		}

		m := client.rpcDialPoolMap[dialPoolGroup]
		entry, ok := m[addr]
		if ok {
			return entry
		}

		entry = &tmcEntry{}
		m[addr] = entry
		return entry
	}()
```

This pattern can be applied to all the other places as well.

Reply:
Sounds like a great approach, incorporated the same.
Thanks a lot for the detailed explanation